# **Model Selection**
This Notebook tests multiple different Machine Learning Algos against our given Problem.


This should give us a first overview and indication towards which models will perform best.


Additionally we will perform Benchmarks, such as RandomGuessing and ZeroGuessing in order to have a real world comparison


The different models we will test are:
- Random Forest
- Decision Tree
- XGBoost 
- LightGBM

## Evaluation Metrics
To evaluate and compare our model we will use typical classification metrics:
- Recall
- Precision
- F1
- ROC-AUC (chooses indpendent threshold for classification. Simply meassures if true has higher % outcome than false)

# **Load merged Dataset**
this Dataset consists of 3 subests:
- X_train, y_train (70%)
- X_val, y_val (15%)
- X_test, y_test (15%)

In [ ]:
import sys
sys.path.append('..')  
from src.data.loader import Loader
from src.data.splitter import split_data
from src.models.random_guessing import RandomGuessing
from src.models.zero_guessing import ZeroGuessing
from src.evaluation.evaluation_metrics import EvaluationMetrics

# model dependencies:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier


loader = Loader().load()
X_train, X_test, y_train, y_test= split_data(loader.df)

0.574871978821797
0


## Benchmarks
This section will test our Benchmarks 
- RandomGuessing : guesses random number between 0 and 1
- ZeroGuessing : always guesses 0 (no fraud)


based on our evaluation matrics against our given dataset

In [2]:
model = RandomGuessing()
y_pred_test = model.predict_many(X=X_test)
random_guessing_metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
random_guessing_metrics.print_eval_report(headline="RandomGuessing")


RandomGuessing
- Precision: 0.034430232839285285
- Recall: 0.7007874015748031
- F1: 0.06563573091194026
- ROC-AUC: 0.5010378212222


In [3]:
model = ZeroGuessing()
y_pred_test = model.predict_many(X=X_test)
zero_guessing_metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
zero_guessing_metrics.print_eval_report(headline="ZeroGuessing")


ZeroGuessing
- Precision: 0.0
- Recall: 0.0
- F1: 0.0
- ROC-AUC: 0.5


/Users/tomseidel/Desktop/Learning/ML/fraud-detection/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


## Models
This Section will test the listed models from above

### Decision Tree

In [4]:
model = DecisionTreeClassifier(random_state=42).fit(X_train, y_train)
y_pred_test = model.predict_proba(X_test)[:, 1] # return 1d array of probabilities y=1
metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
metrics.print_eval_report(headline="DecisionTreeClassifier")

ValueError: could not convert string to float: 'W'

### Random Forest

In [14]:
model = RandomForestClassifier(random_state=42).fit(X_train, y_train)
y_pred_test = model.predict_proba(X_test)[:, 1]
metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
metrics.print_eval_report(headline="RandomForestClassifier")


RandomForestClassifier
- Precision: 0.8248076052512449
- Recall: 0.5879315908357534
- F1: 0.6865109269027883
- ROC-AUC: 0.924338686732084


In [19]:
print(X_train.shape)

(413378, 445)


### XGBoost

In [18]:
model = XGBClassifier(random_state=42).fit(X_train, y_train, eval_set = [(X_val,y_val)])
y_pred_test = model.predict_proba(X_test)[:, 1]
metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
metrics.print_eval_report(headline="XGBClassifier")

[0]	validation_0-logloss:0.11855
[1]	validation_0-logloss:0.11166
[2]	validation_0-logloss:0.10685
[3]	validation_0-logloss:0.10434
[4]	validation_0-logloss:0.10209
[5]	validation_0-logloss:0.09985
[6]	validation_0-logloss:0.09851
[7]	validation_0-logloss:0.09712
[8]	validation_0-logloss:0.09588
[9]	validation_0-logloss:0.09430
[10]	validation_0-logloss:0.09285
[11]	validation_0-logloss:0.09140
[12]	validation_0-logloss:0.09053
[13]	validation_0-logloss:0.08915
[14]	validation_0-logloss:0.08833
[15]	validation_0-logloss:0.08786
[16]	validation_0-logloss:0.08730
[17]	validation_0-logloss:0.08644
[18]	validation_0-logloss:0.08624
[19]	validation_0-logloss:0.08535
[20]	validation_0-logloss:0.08480
[21]	validation_0-logloss:0.08448
[22]	validation_0-logloss:0.08425
[23]	validation_0-logloss:0.08384
[24]	validation_0-logloss:0.08374
[25]	validation_0-logloss:0.08328
[26]	validation_0-logloss:0.08274
[27]	validation_0-logloss:0.08215
[28]	validation_0-logloss:0.08157
[29]	validation_0-loglos

### LightGBM

In [21]:
model = LGBMClassifier(random_state=42).fit(X_train, y_train, eval_set = [(X_val,y_val)])
y_pred_test = model.predict_proba(X_test)[:, 1]
metrics = EvaluationMetrics(y=y_test, y_pred=y_pred_test)
metrics.print_eval_report(headline="LGBMClassifier")

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 14464, number of negative: 398914
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.101557 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 39174
[LightGBM] [Info] Number of data points in the train set: 413378, number of used features: 443
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.034990 -> initscore=-3.317083
[LightGBM] [Info] Start training from score -3.317083

LGBMClassifier
- Precision: 0.7885338345864662
- Recall: 0.5414649887060342
- F1: 0.6420508896116319
- ROC-AUC: 0.925294125276768


# Summary

All four models were tested on raw, unencoded data (no feature engineering), which sets a performance floor for what the final pipeline needs to beat.

**Decision Tree** failed outright — sklearn's implementation requires numeric input and cannot handle raw string columns. This alone ruled it out.

**Random Forest** and **LightGBM** achieved a ROC-AUC (~0.924–0.925).

**XGBoost** was selected for further development because it handles mixed data types and NaN values natively, which makes sense in a setup with very much missing values